src/util/locktime.rs

```diff
@@ -0,0 +1,233 @@
+pub struct Height(u32);
+
+impl Height {
+    /// Constructs a new block height.
+    ///
+    /// # Errors
+    /// If `n` does not encode a block height value (see [`encodes_height`]).
+    pub fn new(n: u32) -> Result<Self, Error> {
+        if encodes_height(n) {
+            Ok(Self(n))
+        } else {
+            Err(Error::InvalidHeight)
+        }
+    }
+
+    /// Converts `Height` to its inner `u32` value.
+    pub fn to_u32(&self) -> u32 {
+        self.0
+    }
+}
+
+/// A UNIX timestamp, seconds since epoch.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Time(u32);
+
+impl Time {
+    /// Constructs a new block time.
+    ///
+    /// # Errors
+    /// If `n` does not encode a UNIX time stamp (see [`encodes_time`]).
+    pub fn new(n: u32) -> Result<Self, Error> {
+        if encodes_time(n) {
+            Ok(Self(n))
+        } else {
+            Err(Error::InvalidTime)
+        }
+    }
+
+    /// Converts `Time` to its inner `u32` value.
+    pub fn to_u32(&self) -> u32 {
+        self.0
+    }
+}
+
+/// Returns true if `n` encodes a block height i.e., is < 500,000,000.
+pub fn encodes_height(n: u32) -> bool {
+    n < LOCKTIME_THRESHOLD
+}
+
+/// Returns true if `n` encodes a UNIX timestamp i.e., is >= 500,000,000.
+pub fn encodes_time(n: u32) -> bool {
+    n >= LOCKTIME_THRESHOLD
+}
+
+/// Returns true if the timelock constraint of `n OP_CHECKLOCKTIMEVERIFY` and nLockTime is
+/// satisfied.
+///
+/// # Errors
+///
+/// If `n` and `n_lock_time` are different units i.e., one represents a block height and the other
+/// represents a block time.
+pub fn is_satisfied(n: LockTime, n_lock_time: LockTime) -> Result<bool, Error> {
+    if !n.is_same_unit(n_lock_time) {
+        return Err(Error::InvalidComparison);
+    }
+    Ok(n_lock_time.to_u32() >= n.to_u32())
+}
+
+impl fmt::Display for LockTime {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            LockTime::Blocks(ref t) => write!(f, "{}", t.0),
+            LockTime::Seconds(ref t) => write!(f, "{}", t.0),
+        }
+    }
+}
+
+impl Encodable for LockTime {
+    fn consensus_encode<W: io::Write>(&self, e: W) -> Result<usize, io::Error> {
+        let v = self.to_u32();
+        v.consensus_encode(e)
+    }
+}
+
+impl Decodable for LockTime {
+    #[inline]
+    fn consensus_decode<D: io::Read>(d: D) -> Result<Self, encode::Error> {
+        let v = u32::consensus_decode(d)?;
+        Ok(LockTime::from(v))
+    }
+}
+
+/// Timelock related errors.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// Cannot compare different lock units (height vs time).
+    InvalidComparison,
+    /// Value does not encode a block height.
+    InvalidHeight,
+    /// Value does not encode a UNIX timestamp.
+    InvalidTime,
+}
```

Comment:
Perhaps separating conversion errors from operation errors would be cleaner?

```rust
// internal, used just for displaying
enum Which {
    Seconds,
    Blocks,
}

pub struct ConversionError {
    // fields intentionally private
    which: Which,
    input: u32,
}

impl fmt::Display for ConversionError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self.which {
            Which::Blocks => write!(f, "invalid number of blocks {} in lock time, must be below 500 million", self.input),
            Which::Seconds => write!(f, "invalid UNIX timestamp {} in lock time, must be at least 500 million seconds", self.input),
        }
    }
}

#[non_exhaustive]
pub enum OperationError {
    InvalidComparison,
}

// catch-all type for errors related to time locks
#[non_exhaustive]
pub enum Error {
    Conversion(ConversionError),
    Operation(OperationError),
}

// impl From<{Operation,Conversion}Error for Error ...
```

Reply:
`s/Which/TimeLockUnit/` or something

Reply:
Thanks for writing out the suggestion, seems over engineered? And `ConversionError` begs the question, "why use a struct with nested `Which` for the error but use an enum for `LockTime`?"

Reply:
Cool, thanks for the explanation, I'll defer to you on the debugging value. I don't understand what you mean by 'conservative', I get the invariant thing though, and that answers my question. I've implemented the error logic as suggested. Thanks for your patience.

Reply:
Conservative API means we don't promise too much so that we don't have to issue breaking version if we decide to change things.

I don't wish for it to happen to you but if you ever encounter code where someone wrote `File::open(...)?` without second thought and then you're wondering which file (out of many the application tried to open) doesn't exist, you will quickly see why having lot of information in error is valuable. :)